### ISSUE TITLE
[3.3.5] Autokick with LuaUnlockers

### ISSUE THREAD

**Issue by Aknunx** [TrinityCore/TrinityCore#17484]
**Description**: You can use Lua Unlocker programs to execute protected functions that shouldn't be avalaible, like http://wowprogramming.com/docs/api/CastSpellByName

**Current behaviour**: Using addons or external programs you're able to create lua scripts to automatize multiple ingame functions, such as kick instantly whenever your target/focus/whatever is casting a spell, by calling protected Blizzard UI functions such as "CastSpell", "CastSpellByName", "CastSpellByID".

**Expected behaviour**: Those lua scripts shouldn't be able to call protected functions.

**Steps to reproduce the problem**:

1.  Download https://mega.nz/#!bp51zRYC!_lBtpfQ6sr7SWQSv3xE7U9gWqqIJryEnpFZsW937qzc and extract it
2. With wow closed, execute the .exe, select the Process "Edit Mode" and press Select
3. Press "Ability And Rotation Editor"
4. On the new tab, select mage, for example, then press "New Profile", introduce a profile name, do the same with "New Rotation" and close the tab.
5. Back in the main menu press "Ability Editor"
6. Again, select our class, mage, type any name on the "Name" field and any id on the "Spell Id" field.
7. Then in the "Test" tab copy our lua script and press "Save":
if ((select(2, GetSpellCooldown("Counterspell")))==0) and (UnitCastingInfo("focus") or UnitChannelInfo("focus")) then CastSpellByName("Counterspell","focus") end
8. Go back to the "Ability and Rotation Editor", select mage and the profile that we've created, our spell should now appear under "Avalaible Abilities".
Press "->" and uncheck "Require combat to execute rotation" box.
9. Close the tab and close the entire program.
10. Execute your wow client, log in with a mage and /focus someone.
11. Execute again the PQR .exe, instead of "Edit Mode" from step 2 you'll see your character profile, select it.
12. Deploy rotation 1 and select your spell profile.
13. Go back to the wow client and do the rotation 1 (alt X), a message will appear telling you that PQR is now active, whenever your /focus casts something you'll instantly counterspell it.

**Branch(es)**: 3.3.5

**TC hash/commit**:  https://github.com/TrinityCore/TrinityCore/commit/3213e4712c75dc09f44bc7dc3fca32efb0f94bc1

**TDB version**:  335.61

**Operating system**:  Ubuntu

[//]: # (This template is for problem reports, for other type of reports edit it accordingly)
[//]: # (If this is a crash report, include the crashlog with https://gist.github.com/)
[//]: # (For fixes containing c++ create a Pull Request)

**Comment by Nawuko** [comment]
First of all there is nothing that TrinityCore can do against client modifications. Second of all this is "blizzlike". Blizzard had struggles with AutoKick scripts in the 3.3.5 area.

The only thing you can really do is to compare the spell cast time with cast time of the counter spell. If its often very small the player is either really good or some kind of script is used. But this is, to my understanding, outside of TrinityCores scope.

Maybe warden can help in this situation. (But i'm almost sure warden has already a offset check for protected lua).

**Comment by Keader** [comment]
Has no way to detect this "unlockers"?

**Comment by Romain-P** [comment]
Nawuko said the way. Create stats for each player and see if the player is frequently casting fast

**Comment by DDuarte** [comment]
I partially agree with Nawuko and I'm tempted to close this as "won't fix".

> First of all there is nothing that TrinityCore can do against client modifications. 

We do can sanitise inputs server side and that is enough against most exploits, however not this one in particular.

**Comment by Raydor** [comment]
I know this is a closed issue but http://forum.wowcircle.com/en/showthread.php?t=5586 these guys managed to made a fix for this without making the player download anything. Just wanted to say that this is not impossible but might be tremendously hard.

**Comment by Nawuko** [comment]
No, the dont make a "fix". They just queried their database to find commonly used macros for that kind of abuse (autokick for example). And i would argue something like this is still not in the scope of TC

**Comment by robinsch** [comment]
> They just queried their database

select hacker from hackerList where type = lua

**Comment by Raydor** [comment]
@Nawuko I know its out of TC scope, antihack (aka warden) should have been out of TC scope... and no, a friend told me that to just test it, he openned PQR and implemented a hack and just at the start of a duel he got banned.

**Comment by robinsch** [comment]
Just check time between cast started and interrupted, if a player kicks multiple times directly on next tick then ban him.

**Comment by Raydor** [comment]
@robinsch Its not only with autokick but with spell rotations and all other PQR hacks, they just detect the PQR program.

**Comment by robinsch** [comment]
I know, there is always a way to try to filter out potential hackers, rotation could be checked by GCD fade time and spell start time. since they spam abilitys.

**Comment by Nawuko** [comment]
I would argue that Warden is in the scope of TC. TC is first-hand a research/learn project. So reverse engineering the communication between the client anti-cheat and the server part would be fit into the scope of TC. Scanning for LUA strings in macros is not. But there are some usefull points where to start in this conversation, look at the comments from @robinsch, @Nawuko

**Comment by P-Kito** [comment]
What about something like this?
Save kick times and compare them if it is    +/- 50ms at the same time for X times = Log it.

--- PSEUDOCODE ---

uint8 threshold = 5;

uint32 kickTime = castTime - interruptTime;
if (oldKickTime == kickTime)
    m_sameKickTime++;

if (m_sameKickTime >= threshold)
   LOG("player kicked more than %u times at the same time, banning", threshold);


--- PSEUDOCODE ---

**Comment by tkrokli** [comment]
This discussion has become quite meta (as it is not a TrinityCore issue). What about using the Forum?

**Comment by Laintime** [comment]
Why this closing?

**Comment by tkrokli** [comment]
@Laintime : Because the discussion is not about the TrinityCore source, it is about Lua and other things way outside of TrinityCore.
